// vocs/_lib/app/components/SearchDialog.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as Dialog from '@radix-ui/react-dialog';
import { ArrowLeftIcon, ChevronRightIcon, FileIcon, ListBulletIcon, MagnifyingGlassIcon, } from '@radix-ui/react-icons';
import * as Label from '@radix-ui/react-label';
import clsx from 'clsx';
import { default as Mark } from 'mark.js';
import {} from 'minisearch';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { useConfig } from '../hooks/useConfig.js';
import { useDebounce } from '../hooks/useDebounce.js';
import { useLocalStorage } from '../hooks/useLocalStorage.js';
import { useSearchIndex } from '../hooks/useSearchIndex.js';
import { visuallyHidden } from '../styles/utils.css.js';
import { Content } from './Content.js';
import { KeyboardShortcut } from './KeyboardShortcut.js';
import * as styles from './SearchDialog.css.js';
export function SearchDialog(props) {
    const { search: searchOptions } = useConfig();
    const navigate = useNavigate();
    const inputRef = useRef(null);
    const listRef = useRef(null);
    const [filterText, setFilterText] = useLocalStorage('filterText', '');
    const searchTerm = useDebounce(filterText, 200);
    const searchIndex = useSearchIndex();
    const [selectedIndex, setSelectedIndex] = useState(-1);
    const [disableMouseOver, setDisableMouseOver] = useState(false);
    const [showDetailView, setShowDetailView] = useLocalStorage('showDetailView', true);
    const results = useMemo(() => {
        if (!searchIndex)
            return [];
        if (!searchTerm) {
            setSelectedIndex(-1);
            return [];
        }
        setSelectedIndex(0);
        return searchIndex.search(searchTerm, searchOptions).slice(0, 16);
    }, [searchIndex, searchOptions, searchTerm]);
    const resultsCount = results.length;
    const selectedResult = results[selectedIndex];
    const highlight = useCallback(() => {
        if (!listRef.current)
            return;
        const terms = new Set();
        for (const result of results) {
            for (const term in result.match) {
                terms.add(term);
            }
        }
        const mark = new Mark(listRef.current);
        mark.unmark({
            done() {
                mark?.markRegExp(formMarkRegex(terms));
            },
        });
        const excerptElements = listRef.current.querySelectorAll(`.${styles.excerpt}`);
        for (const element of excerptElements) {
            element.querySelector('mark[data-markjs="true"]')?.scrollIntoView({ block: 'center' });
        }
        listRef.current?.firstElementChild?.scrollIntoView({ block: 'start' });
    }, [results]);
    useEffect(() => {
        if (!props.open)
            return;
        function keyDownHandler(event) {
            switch (event.key) {
                case 'ArrowDown': {
                    event.preventDefault();
                    setSelectedIndex((index) => {
                        let nextIndex = index + 1;
                        if (nextIndex >= resultsCount)
                            nextIndex = 0;
                        const element = listRef.current?.children[nextIndex];
                        element?.scrollIntoView({ block: 'nearest' });
                        return nextIndex;
                    });
                    setDisableMouseOver(true);
                    break;
                }
                case 'ArrowUp': {
                    event.preventDefault();
                    setSelectedIndex((index) => {
                        let nextIndex = index - 1;
                        if (nextIndex < 0)
                            nextIndex = resultsCount - 1;
                        const element = listRef.current?.children[nextIndex];
                        element?.scrollIntoView({ block: 'nearest' });
                        return nextIndex;
                    });
                    setDisableMouseOver(true);
                    break;
                }
                case 'Backspace': {
                    if (!event.metaKey)
                        return;
                    event.preventDefault();
                    setFilterText('');
                    inputRef.current?.focus();
                    break;
                }
                case 'Enter': {
                    if (event.target instanceof HTMLButtonElement && event.target.type !== 'submit')
                        return;
                    if (!selectedResult)
                        return;
                    event.preventDefault();
                    navigate(selectedResult.href);
                    props.onClose();
                    break;
                }
            }
        }
        window.addEventListener('keydown', keyDownHandler);
        return () => {
            window.removeEventListener('keydown', keyDownHandler);
        };
    }, [navigate, resultsCount, setFilterText, selectedResult, props.open, props.onClose]);
    useEffect(() => {
        if (searchTerm === '')
            return;
        if (!listRef.current)
            return;
        highlight();
    }, [highlight, searchTerm]);
    return (_jsxs(Dialog.Portal, { children: [_jsx(Dialog.Overlay, { className: styles.overlay }), _jsxs(Dialog.Content, { onOpenAutoFocus: (event) => {
                    if (inputRef.current) {
                        event.preventDefault();
                        inputRef.current.focus();
                    }
                    highlight();
                }, onCloseAutoFocus: () => {
                    setSelectedIndex(0);
                }, className: styles.root, "aria-describedby": undefined, children: [_jsx(Dialog.Title, { className: visuallyHidden, children: "Search" }), _jsxs("form", { className: styles.searchBox, children: [_jsx("button", { "aria-label": "Close search dialog", type: "button", onClick: () => props.onClose(), className: styles.searchInputIconMobile, children: _jsx(ArrowLeftIcon, { className: styles.searchInputIcon, height: 20, width: 20 }) }), _jsx(Label.Root, { htmlFor: "search-input", children: _jsx(MagnifyingGlassIcon, { "aria-label": "Search", className: clsx(styles.searchInputIcon, styles.searchInputIconDesktop), height: 20, width: 20 }) }), _jsx("input", { ref: inputRef, tabIndex: 0, className: styles.searchInput, id: "search-input", onChange: (event) => setFilterText(event.target.value), placeholder: "Search", type: "search", value: filterText }), _jsx("button", { "aria-label": "Toggle detail view", type: "button", onClick: () => setShowDetailView((x) => !x), children: _jsx(ListBulletIcon, { className: styles.searchInputIcon, height: 20, width: 20 }) }), _jsx("button", { "aria-label": "Reset search", type: "button", className: styles.searchInputIcon, onClick: () => {
                                    setFilterText('');
                                    inputRef.current?.focus();
                                }, children: "\u232B" })] }), _jsxs("ul", { className: styles.results, role: results.length ? 'listbox' : undefined, onMouseMove: () => setDisableMouseOver(false), ref: listRef, children: [searchTerm && results.length === 0 && (_jsxs("li", { children: ["No results for \"", _jsx("span", { children: searchTerm }), "\""] })), results.map((result, index) => (_jsx("li", { 
                                // biome-ignore lint/a11y/noNoninteractiveElementToInteractiveRole:
                                role: "option", className: clsx(styles.result, index === selectedIndex && styles.resultSelected), "aria-selected": index === selectedIndex, "aria-label": [...result.titles.filter((title) => Boolean(title)), result.title].join(' > '), children: _jsxs(Link, { to: result.href, onClick: (event) => {
                                        // Don't close the dialog if the user is opening the link in a new tab.
                                        if (event.metaKey)
                                            return;
                                        props.onClose();
                                    }, onMouseEnter: () => !disableMouseOver && setSelectedIndex(index), onFocus: () => setSelectedIndex(index), children: [_jsxs("div", { className: styles.titles, children: [result.isPage ? (_jsx(FileIcon, { className: styles.resultIcon })) : (_jsx("span", { className: styles.resultIcon, children: "#" })), result.titles
                                                    .filter((title) => Boolean(title))
                                                    .map((title) => (_jsxs("span", { className: styles.title, children: [_jsx("span", { 
                                                            // biome-ignore lint/security/noDangerouslySetInnerHtml:
                                                            dangerouslySetInnerHTML: { __html: title } }), _jsx(ChevronRightIcon, { className: styles.titleIcon })] }, title))), _jsx("span", { className: styles.title, children: _jsx("span", { 
                                                        // biome-ignore lint/security/noDangerouslySetInnerHtml:
                                                        dangerouslySetInnerHTML: { __html: result.title } }) })] }), showDetailView && result.text?.trim() && (_jsx("div", { className: styles.excerpt, children: _jsx(Content, { className: styles.content, children: _jsx("div", { 
                                                    // biome-ignore lint/security/noDangerouslySetInnerHtml:
                                                    dangerouslySetInnerHTML: { __html: result.html } }) }) }))] }) }, result.id)))] }), _jsxs("div", { className: styles.searchShortcuts, children: [_jsx(KeyboardShortcut, { description: "Navigate", keys: ['↑', '↓'] }), _jsx(KeyboardShortcut, { description: "Select", keys: ['enter'] }), _jsx(KeyboardShortcut, { description: "Close", keys: ['esc'] }), _jsx(KeyboardShortcut, { description: "Reset", keys: ['⌘', '⌫'] })] })] })] }));
}
function formMarkRegex(terms) {
    return new RegExp([...terms]
        .sort((a, b) => b.length - a.length)
        .map((term) => {
        return `(${term.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&').replace(/-/g, '\\x2d')})`;
    })
        .join('|'), 'gi');
}
